src/sentry/auth/helper.py

```diff
@@ -114,415 +117,443 @@ def __setattr__(self, key, value):
-    # we trust identity providers, its considered safe.
-    # Note: we do not trust things like SAML, so the SSO implementation needs
-    # to consider if 'email_verified' can be trusted or not
-    if acting_user and identity.get("email_verified"):
-        # we only allow this flow to happen if the existing user has
-        # membership, otherwise we short circuit because it might be
-        # an attempt to hijack membership of another organization
-        has_membership = OrganizationMember.objects.filter(
-            user=acting_user, organization=organization
-        ).exists()
-        if has_membership:
-            if not auth.login(
-                request,
-                acting_user,
-                after_2fa=request.build_absolute_uri(),
-                organization_id=organization.id,
-            ):
-                if acting_user.has_usable_password():
-                    return post_login_redirect(request)
+        - Is there an existing user (via authentication) that should be merged?
+
+        - Should I create a new user based on this identity?
+        """
+        op = self.request.POST.get("op")
+        if not self.user.is_authenticated:
+            # TODO(dcramer): its possible they have multiple accounts and at
+            # least one is managed (per the check below)
+            try:
+                acting_user = User.objects.filter(
+                    id__in=UserEmail.objects.filter(email__iexact=identity["email"]).values("user"),
+                    is_active=True,
+                ).first()
+            except IndexError:
+                acting_user = None
+            login_form = AuthenticationForm(
+                self.request,
+                self.request.POST if self.request.POST.get("op") == "login" else None,
+                initial={"username": acting_user.username if acting_user else None},
+            )
+        else:
+            acting_user = self.user
+
+        # If they already have an SSO account and the identity provider says
+        # the email matches we go ahead and let them merge it. This is the
+        # only way to prevent them having duplicate accounts, and because
+        # we trust identity providers, its considered safe.
+        # Note: we do not trust things like SAML, so the SSO implementation needs
+        # to consider if 'email_verified' can be trusted or not
+        if acting_user and identity.get("email_verified"):
+            # we only allow this flow to happen if the existing user has
+            # membership, otherwise we short circuit because it might be
+            # an attempt to hijack membership of another organization
+            has_membership = OrganizationMember.objects.filter(
+                user=acting_user, organization=self.organization
+            ).exists()
+            if has_membership:
+                try:
+                    self._login(acting_user)
+                except self._NotLoggedIn:
```

Comment:
this pattern of wrapping login with an exception feels weird to me -- would the ideal solution be to raise an exception within utils.login itself? i see that we don't really use the value returned from this function anywhere but here (guessing we rely on the side affects of login not completing to accomplish correct flow elsewhere).

it also seems like the only way login would return False is if the user has 2FA and hasn't completed it -- would `NotCompletedSecurityChecks` be a better fitting exception name?

Reply:
> this pattern of wrapping login with an exception feels weird to me

It feels weird to me too in general, but as I understand it, the Pythonic style is more liberal about exceptions than most other languages. If you're trying to represent an exit condition from a function, an exception is preferred over a return value.

In this particular case, I do like it because most of the `if not auth.login(...):` blocks would interrupt the flow with a `return` statement. Putting flow control in the `except` block feels more natural to me than just any old code.

> would the ideal solution be to raise an exception within utils.login itself?

I would indeed love to make that change. I avoided making a change with such a large impact on other code, but...

> i see that we don't really use the value returned from this function anywhere but here (guessing we rely on the side affects of login not completing to accomplish correct flow elsewhere).

Interesting point. Let me take a look.

> it also seems like the only way login would return False is if the user has 2FA and hasn't completed it -- would `NotCompletedSecurityChecks` be a better fitting exception name?

Nice point. I'll double-check this and make the change.